packages/react-components/react-dialog/Spec.md

```diff
@@ -1,63 +1,459 @@
+
+```jsx
+<Dialog
+  cancelButton="Connect protocol"
+  confirmButton="Transmit capacitor"
+  content="Connect driver"
+  header="Transmit capacitor"
+  headerAction="Generate protocol"
+  trigger={<Button content="A trigger" />}
+/>
+```
 
 ## API
 
-_List the **Props** and **Slots** proposed for the component. Ideally this would just be a link to the component's `.types.ts` file_
+The `Dialog` should implement a `children` based API as is one of the standards across all the surveyed alternatives as a part of Open UI research in [Prior Art](#prior-art). The component will leverage the use of `context` in the interaction and data flows of child compound components.
 
-## Structure
+Sample usages will be give in the following section of this document [Sample code](#sample-code)
 
-- _**Public**_
-- _**Internal**_
-- _**DOM** - how the component will be rendered as HTML elements_
+### Dialog
 
-## Migration
+The root level component serves as an interface for interaction with all possible behaviors exposed. It provides context down the hierarchy to `children` compound components to allow functionality.
+
+```tsx
+type DialogSlots = {
+  /**
+   * The dialog element itself
+   */
+  root: Slot<'div'>;
+  /**
+   * Dimmed background of dialog.
+   * The default overlay is rendered as a `<div>` with styling.
+   * This slot expects a `<div>` element which will replace the default overlay.
+   * The overlay should have `aria-hidden="true"`.
+   */
+  overlay?: Slot<'div'>;
+};
+
+type DialogProps = ComponentProps<DialogSlots> & {
+  /**
+   * Dialog variations.
+   *
+   * `modal`: When this type of dialog is open, the rest of the page is dimmed out and cannot be interacted with. The tab sequence is kept within the dialog and moving the focus outside the dialog will imply closing it. This is the default type of the component.
+   *
+   * `non-modal`: When a non-modal dialog is open, the rest of the page is not dimmed out and users can interact with the rest of the page. This also implies that the tab focus can move outside the dialog when it reaches the last focusable element.
+   *
+   * `alert`: is a special type of modal dialogs that interrupts the user's workflow to communicate an important message or ask for a decision. Unlike a typical modal dialog, the user must take an action through the options given to dismiss the dialog, and it cannot be dismissed through the dimmed background or escape key.
+   */
+  type?: 'modal' | 'non-modal' | 'alert';
+  /**
+   * Controls the open state of the dialog
+   * @default undefined
+   */
+  open?: boolean;
+  /**
+   * Default value for the uncontrolled open state of the dialog.
+   * @default false
+   */
+  defaultOpen?: boolean;
+  /**
+   * Callback fired when the component requests to be closed.
+   * @default undefined
+   */
+  onClose?(event: MouseEvent | KeyboardEvent, data: DialogCloseData): void;
```

Comment:
how would we differentiate between cancel and confirm?

Reply:
`cancel` or `confirm` are formularies states, not a dialog problem per se.

Adding a `onClick` handler on the trigger element should be enough to capture those states, we could even implement `event.preventDefault` to prevent the dialog from closing on the case that a formulary is not correctly filled.

Reply:
a dialog with cancel and confirm is very common - can you please create a prototype of it with the proposed approach so that we would see how easy/hard would it be to handle all the cases correclty (button click, outside click, esc)

Reply:
Yes, I'll be adding a sample code of usage with formularies with controls and without!

Reply:
@jurokapsiar, see [Async confirm dialog example](https://github.com/microsoft/fluentui/blob/02dca57fac036f2076f3395a959e5f408ad99ace/packages/react-components/react-dialog/Spec.md#async-confirm-dialog)

Reply:
I don't think the above example handles all - what I meant is a dialog that for example does:
- sending data on confirm
- cleanup on cancel
I would be interested to see how much does the consumer need to do to distinguish between these two states in the cancel and confirm handlers

Reply:
Ok, I'll add another example for that also